File: backend/src/aexy/models/app_definitions.py
"""App definitions catalog with modules and bundle templates.

This module defines all available apps in the system, their sub-modules,
required permissions, and pre-configured access bundle templates.
"""

from enum import Enum
from typing import TypedDict


class AppCategory(str, Enum):
    """App categories for grouping in UI."""

    ENGINEERING = "engineering"
    PEOPLE = "people"
    BUSINESS = "business"
    PRODUCTIVITY = "productivity"


class ModuleConfig(TypedDict, total=False):
    """Configuration for an app module."""

    name: str
    description: str
    route: str  # Relative route within the app


class AppConfig(TypedDict, total=False):
    """Configuration for an app."""

    name: str
    description: str
    icon: str
    category: AppCategory
    base_route: str
    required_permission: str | None  # None means accessible to all
    modules: dict[str, ModuleConfig]


# Master app catalog defining all apps and their modules
APP_CATALOG: dict[str, AppConfig] = {
    "dashboard": {
        "name": "Dashboard",
        "description": "Overview and analytics dashboard",
        "icon": "LayoutDashboard",
        "category": AppCategory.PRODUCTIVITY,
        "base_route": "/dashboard",
        "required_permission": None,  # Accessible to all authenticated users
        "modules": {},
    },
    "tracking": {
        "name": "Tracking",
        "description": "Standups, blockers, and time tracking",
        "icon": "Activity",
        "category": AppCategory.ENGINEERING,
        "base_route": "/tracking",
        "required_permission": "can_view_tracking",
        "modules": {
            "standups": {
                "name": "Standups",
                "description": "Daily standup submissions and history",
                "route": "/standups",
            },
            "blockers": {
                "name": "Blockers",
                "description": "Track and manage blockers",
                "route": "/blockers",
            },
            "time": {
                "name": "Time Tracking",
                "description": "Log and track work hours",
                "route": "/time",
            },
        },
    },
    "sprints": {
        "name": "Sprints",
        "description": "Sprint planning and task management",
        "icon": "Zap",
        "category": AppCategory.ENGINEERING,
        "base_route": "/sprints",
        "required_permission": "can_view_sprints",
        "modules": {
            "board": {
                "name": "Sprint Board",
                "description": "Kanban-style sprint board",
                "route": "/board",
            },
            "epics": {
                "name": "Epics",
                "description": "Manage epics and user stories",
                "route": "/epics",
            },
            "tasks": {
                "name": "Tasks",
                "description": "Task management and assignment",
                "route": "/tasks",
            },
            "backlog": {
                "name": "Backlog",
                "description": "Product backlog management",
                "route": "/backlog",
            },
        },
    },
    "tickets": {
        "name": "Tickets",
        "description": "Support ticket management",
        "icon": "Ticket",
        "category": AppCategory.BUSINESS,
        "base_route": "/tickets",
        "required_permission": "can_view_tickets",
        "modules": {},
    },
    "reviews": {
        "name": "Reviews",
        "description": "Performance reviews and feedback",
        "icon": "Star",
        "category": AppCategory.PEOPLE,
        "base_route": "/reviews",
        "required_permission": "can_view_reviews",
        "modules": {
            "cycles": {
                "name": "Review Cycles",
                "description": "Review cycle management",
                "route": "/cycles",
            },
            "goals": {
                "name": "Goals",
                "description": "Work goals and OKRs",
                "route": "/goals",
            },
            "peer_requests": {
                "name": "Peer Requests",
                "description": "Peer feedback requests",
                "route": "/peer-requests",
            },
            "manage": {
                "name": "Manage",
                "description": "Admin review management",
                "route": "/manage",
            },
        },
    },
    "hiring": {
        "name": "Hiring",
        "description": "Recruitment and assessments",
        "icon": "Users",
        "category": AppCategory.PEOPLE,
        "base_route": "/hiring",
        "required_permission": "can_view_hiring",
        "modules": {
            "dashboard": {
                "name": "Dashboard",
                "description": "Hiring overview and metrics",
                "route": "/dashboard",
            },
            "candidates": {
                "name": "Candidates",
                "description": "Manage candidates",
                "route": "/candidates",
            },
            "assessments": {
                "name": "Assessments",
                "description": "Technical assessments",
                "route": "/assessments",
            },
            "questions": {
                "name": "Question Bank",
                "description": "Assessment questions library",
                "route": "/questions",
            },
            "templates": {
                "name": "Templates",
                "description": "Assessment templates",
                "route": "/templates",
            },
            "analytics": {
                "name": "Analytics",
                "description": "Hiring analytics and reports",
                "route": "/analytics",
            },
        },
    },
    "learning": {
        "name": "Learning",
        "description": "Learning paths and courses",
        "icon": "GraduationCap",
        "category": AppCategory.PEOPLE,
        "base_route": "/learning",
        "required_permission": "can_view_learning",
        "modules": {},
    },
    "crm": {
        "name": "CRM",
        "description": "Customer relationship management",
        "icon": "Building2",
        "category": AppCategory.BUSINESS,
        "base_route": "/crm",
        "required_permission": "can_view_crm",
        "modules": {
            "overview": {
                "name": "Overview",
                "description": "CRM dashboard and pipeline",
                "route": "/overview",
            },
            "inbox": {
                "name": "Inbox",
                "description": "Email inbox and communications",
                "route": "/inbox",
            },
            "agents": {
                "name": "AI Agents",
                "description": "Configure AI sales agents",
                "route": "/agents",
            },
            "activities": {
                "name": "Activities",
                "description": "Activity tracking and logs",
                "route": "/activities",
            },
            "automations": {
                "name": "Automations",
                "description": "Sales automations and sequences",
                "route": "/automations",
            },
            "calendar": {
                "name": "Calendar",
                "description": "Meeting and event calendar",
                "route": "/calendar",
            },
        },
    },
    "email_marketing": {
        "name": "Email Marketing",
        "description": "Email campaigns and automation",
        "icon": "Mail",
        "category": AppCategory.BUSINESS,
        "base_route": "/email-marketing",
        "required_permission": "can_view_crm",  # Uses same permission as CRM
        "modules": {
            "campaigns": {
                "name": "Campaigns",
                "description": "Email campaign management",
                "route": "/campaigns",
            },
            "templates": {
                "name": "Templates",
                "description": "Email templates library",
                "route": "/templates",
            },
            "settings": {
                "name": "Settings",
                "description": "Email settings and domains",
                "route": "/settings",
            },
        },
    },
    "docs": {
        "name": "Docs",
        "description": "Documentation and wiki",
        "icon": "FileText",
        "category": AppCategory.PRODUCTIVITY,
        "base_route": "/docs",
        "required_permission": "can_view_docs",
        "modules": {},
    },
    "forms": {
        "name": "Forms",
        "description": "Form builder and submissions",
        "icon": "ClipboardList",
        "category": AppCategory.PRODUCTIVITY,
        "base_route": "/forms",
        "required_permission": "can_view_forms",
        "modules": {},
    },
    "oncall": {
        "name": "On-Call",
        "description": "On-call schedules and rotations",
        "icon": "Phone",
        "category": AppCategory.ENGINEERING,
        "base_route": "/oncall",
        "required_permission": "can_view_oncall",
        "modules": {},
    },
    "compliance": {
        "name": "Compliance",
        "description": "Compliance management, documents, and reminders",
        "icon": "ShieldCheck",
        "category": AppCategory.PEOPLE,
        "base_route": "/compliance",
        "required_permission": "can_view_compliance",
        "modules": {
            "reminders": {
                "name": "Reminders",
                "description": "Recurring compliance reminders",
                "route": "/reminders",
            },
            "document_center": {
                "name": "Document Center",
                "description": "Upload and manage compliance documents",
                "route": "/documents",
            },
            "training": {
                "name": "Training",
                "description": "Mandatory training management",
                "route": "/training",
            },
            "certifications": {
                "name": "Certifications",
                "description": "Certification tracking",
                "route": "/certifications",
            },
        },
    },
}


class BundleConfig(TypedDict, total=False):
    """Configuration for an app bundle template."""

    name: str
    description: str
    icon: str
    color: str
    apps: dict[str, dict]  # App ID -> {"enabled": bool, "modules": {module_id: bool}}


# System app bundle templates
SYSTEM_APP_BUNDLES: dict[str, BundleConfig] = {
    "engineering": {
        "name": "Engineering",
        "description": "Apps for software development teams",
        "icon": "Code",
        "color": "#2563eb",  # blue
        "apps": {
            "dashboard": {"enabled": True, "modules": {}},
            "tracking": {
                "enabled": True,
                "modules": {"standups": True, "blockers": True, "time": True},
            },
            "sprints": {
                "enabled": True,
                "modules": {
                    "board": True,
                    "epics": True,
                    "tasks": True,
                    "backlog": True,
                },
            },
            "tickets": {"enabled": True, "modules": {}},
            "docs": {"enabled": True, "modules": {}},
            "learning": {"enabled": True, "modules": {}},
            "oncall": {"enabled": True, "modules": {}},
            # Disabled for engineering
            "reviews": {"enabled": False},
            "hiring": {"enabled": False},
            "crm": {"enabled": False},
            "email_marketing": {"enabled": False},
            "forms": {"enabled": False},
            "compliance": {"enabled": False},
        },
    },
    "people": {
        "name": "People",
        "description": "Apps for HR and people operations",
        "icon": "Heart",
        "color": "#f43f5e",  # rose
        "apps": {
            "dashboard": {"enabled": True, "modules": {}},
            "reviews": {
                "enabled": True,
                "modules": {
                    "cycles": True,
                    "goals": True,
                    "peer_requests": True,
                    "manage": True,
                },
            },
            "hiring": {
                "enabled": True,
                "modules": {
                    "dashboard": True,
                    "candidates": True,
                    "assessments": True,
                    "questions": True,
                    "templates": True,
                    "analytics": True,
                },
            },
            "learning": {"enabled": True, "modules": {}},
            "compliance": {
                "enabled": True,
                "modules": {
                    "reminders": True,
                    "document_center": True,
                    "training": True,
                    "certifications": True,
                },
            },
            "docs": {"enabled": True, "modules": {}},
            "forms": {"enabled": True, "modules": {}},
            # Disabled for people ops
            "tracking": {"enabled": False},
            "sprints": {"enabled": False},
            "tickets": {"enabled": False},
            "crm": {"enabled": False},
            "email_marketing": {"enabled": False},
            "oncall": {"enabled": False},
        },
    },
    "business": {
        "name": "Business",
        "description": "Apps for sales and customer success",
        "icon": "Briefcase",
        "color": "#06b6d4",  # cyan
        "apps": {
            "dashboard": {"enabled": True, "modules": {}},
            "crm": {
                "enabled": True,
                "modules": {
                    "overview": True,
                    "inbox": True,
                    "agents": True,
                    "activities": True,
                    "automations": True,
                    "calendar": True,
                },
            },
            "email_marketing": {
                "enabled": True,
                "modules": {"campaigns": True, "templates": True, "settings": True},
            },
            "tickets": {"enabled": True, "modules": {}},
            "docs": {"enabled": True, "modules": {}},
            "forms": {"enabled": True, "modules": {}},
            # Disabled for business
            "tracking": {"enabled": False},
            "sprints": {"enabled": False},
            "reviews": {"enabled": False},
            "hiring": {"enabled": False},
            "learning": {"enabled": False},
            "oncall": {"enabled": False},
            "compliance": {"enabled": False},
        },
    },
    "full_access": {
        "name": "Full Access",
        "description": "Access to all apps and modules",
        "icon": "Shield",
        "color": "#9333ea",  # purple
        "apps": {
            "dashboard": {"enabled": True, "modules": {}},
            "tracking": {
                "enabled": True,
                "modules": {"standups": True, "blockers": True, "time": True},
            },
            "sprints": {
                "enabled": True,
                "modules": {
                    "board": True,
                    "epics": True,
                    "tasks": True,
                    "backlog": True,
                },
            },
            "tickets": {"enabled": True, "modules": {}},
            "reviews": {
                "enabled": True,
                "modules": {
                    "cycles": True,
                    "goals": True,
                    "peer_requests": True,
                    "manage": True,
                },
            },
            "hiring": {
                "enabled": True,
                "modules": {
                    "dashboard": True,
                    "candidates": True,
                    "assessments": True,
                    "questions": True,
                    "templates": True,
                    "analytics": True,
                },
            },
            "learning": {"enabled": True, "modules": {}},
            "crm": {
                "enabled": True,
                "modules": {
                    "overview": True,
                    "inbox": True,
                    "agents": True,
                    "activities": True,
                    "automations": True,
                    "calendar": True,
                },
            },
            "email_marketing": {
                "enabled": True,
                "modules": {"campaigns": True, "templates": True, "settings": True},
            },
            "docs": {"enabled": True, "modules": {}},
            "forms": {"enabled": True, "modules": {}},
            "oncall": {"enabled": True, "modules": {}},
            "compliance": {
                "enabled": True,
                "modules": {
                    "reminders": True,
                    "document_center": True,
                    "training": True,
                    "certifications": True,
                },
            },
        },
    },
}


# Default app access for role templates
# Maps role template ID to a bundle template ID or custom config
ROLE_DEFAULT_APP_ACCESS: dict[str, str] = {
    "owner": "full_access",
    "admin": "full_access",
    "manager": "full_access",
    "developer": "engineering",
    "hr": "people",
    "support": "business",
    "sales": "business",
    "viewer": "engineering",  # Limited view-only access
    "member": "engineering",
}


def get_app_list() -> list[dict]:
    """Get list of all apps with their metadata."""
    return [
        {
            "id": app_id,
            "name": config["name"],
            "description": config["description"],
            "icon": config["icon"],
            "category": config["category"].value,
            "base_route": config["base_route"],
            "required_permission": config.get("required_permission"),
            "modules": [
                {
                    "id": mod_id,
                    "name": mod_config["name"],
                    "description": mod_config["description"],
                    "route": mod_config["route"],
                }
                for mod_id, mod_config in config.get("modules", {}).items()
            ],
        }
        for app_id, config in APP_CATALOG.items()
    ]


def get_bundle_list() -> list[dict]:
    """Get list of all system app bundles."""
    return [
        {
            "id": bundle_id,
            "name": config["name"],
            "description": config["description"],
            "icon": config["icon"],
            "color": config["color"],
            "is_system": True,
            "app_config": config["apps"],
        }
        for bundle_id, config in SYSTEM_APP_BUNDLES.items()
    ]


def get_default_app_access_for_role(role_template_id: str) -> dict:
    """Get default app access config for a role template."""
    bundle_id = ROLE_DEFAULT_APP_ACCESS.get(role_template_id, "engineering")
    bundle = SYSTEM_APP_BUNDLES.get(bundle_id, SYSTEM_APP_BUNDLES["engineering"])
    return bundle["apps"]


def validate_app_access_config(config: dict) -> tuple[bool, str | None]:
    """
    Validate an app access configuration.

    Returns:
        Tuple of (is_valid, error_message)
    """
    if not isinstance(config, dict):
        return False, "App access config must be a dictionary"

    apps = config.get("apps", {})
    if not isinstance(apps, dict):
        return False, "apps field must be a dictionary"

    for app_id, app_config in apps.items():
        if app_id not in APP_CATALOG:
            return False, f"Unknown app: {app_id}"

        if not isinstance(app_config, dict):
            return False, f"Config for {app_id} must be a dictionary"

        if "enabled" not in app_config:
            return False, f"Missing 'enabled' field for app: {app_id}"

        if not isinstance(app_config["enabled"], bool):
            return False, f"'enabled' must be boolean for app: {app_id}"

        # Validate modules if present
        modules = app_config.get("modules", {})
        if modules:
            valid_modules = set(APP_CATALOG[app_id].get("modules", {}).keys())
            for mod_id in modules:
                if mod_id not in valid_modules:
                    return False, f"Unknown module '{mod_id}' for app: {app_id}"

    return True, None

File: frontend/src/config/appDefinitions.ts
/**
 * App Definitions Configuration
 * Defines all available apps, their modules, and route mappings
 * Used for app access control and sidebar filtering
 */

import {
  LayoutDashboard,
  Activity,
  Zap,
  Ticket,
  Star,
  Users,
  GraduationCap,
  Building2,
  Mail,
  FileText,
  ClipboardList,
  Phone,
  CalendarCheck,
  MonitorCheck,
  Bot,
  TrendingUp,
  LucideIcon,
} from "lucide-react";

export type AppCategory = "engineering" | "people" | "business" | "productivity";

export interface AppModule {
  id: string;
  name: string;
  description: string;
  route: string; // Relative route within the app
}

export interface AppDefinition {
  id: string;
  name: string;
  description: string;
  icon: LucideIcon;
  category: AppCategory;
  baseRoute: string;
  requiredPermission: string | null;
  modules: AppModule[];
}

export interface AppBundleTemplate {
  id: string;
  name: string;
  description: string;
  icon: string;
  color: string;
  isSystem: boolean;
  appConfig: Record<string, AppAccessConfig>;
}

export interface AppAccessConfig {
  enabled: boolean;
  modules?: Record<string, boolean>;
}

export interface EffectiveAppAccess {
  appId: string;
  enabled: boolean;
  modules: Record<string, boolean>;
}

export interface MemberAppAccess {
  apps: Record<string, EffectiveAppAccess>;
  appliedTemplateId: string | null;
  appliedTemplateName: string | null;
  hasCustomOverrides: boolean;
  isAdmin: boolean;
}

// Master app catalog
export const APP_CATALOG: Record<string, AppDefinition> = {
  dashboard: {
    id: "dashboard",
    name: "Dashboard",
    description: "Overview and analytics dashboard",
    icon: LayoutDashboard,
    category: "productivity",
    baseRoute: "/dashboard",
    requiredPermission: null,
    modules: [],
  },
  tracking: {
    id: "tracking",
    name: "Tracking",
    description: "Standups, blockers, and time tracking",
    icon: Activity,
    category: "engineering",
    baseRoute: "/tracking",
    requiredPermission: "can_view_tracking",
    modules: [
      { id: "standups", name: "Standups", description: "Daily standup submissions", route: "/standups" },
      { id: "blockers", name: "Blockers", description: "Track and manage blockers", route: "/blockers" },
      { id: "time", name: "Time Tracking", description: "Log and track work hours", route: "/time" },
    ],
  },
  sprints: {
    id: "sprints",
    name: "Sprints",
    description: "Sprint planning and task management",
    icon: Zap,
    category: "engineering",
    baseRoute: "/sprints",
    requiredPermission: "can_view_sprints",
    modules: [
      { id: "board", name: "Sprint Board", description: "Kanban-style sprint board", route: "/board" },
      { id: "epics", name: "Epics", description: "Manage epics and user stories", route: "/epics" },
      { id: "tasks", name: "Tasks", description: "Task management and assignment", route: "/tasks" },
      { id: "backlog", name: "Backlog", description: "Product backlog management", route: "/backlog" },
    ],
  },
  tickets: {
    id: "tickets",
    name: "Tickets",
    description: "Support ticket management",
    icon: Ticket,
    category: "business",
    baseRoute: "/tickets",
    requiredPermission: "can_view_tickets",
    modules: [],
  },
  reviews: {
    id: "reviews",
    name: "Reviews",
    description: "Performance reviews and feedback",
    icon: Star,
    category: "people",
    baseRoute: "/reviews",
    requiredPermission: "can_view_reviews",
    modules: [
      { id: "cycles", name: "Review Cycles", description: "Review cycle management", route: "/cycles" },
      { id: "goals", name: "Goals", description: "Work goals and OKRs", route: "/goals" },
      { id: "peer_requests", name: "Peer Requests", description: "Peer feedback requests", route: "/peer-requests" },
      { id: "manage", name: "Manage", description: "Admin review management", route: "/manage" },
    ],
  },
  hiring: {
    id: "hiring",
    name: "Hiring",
    description: "Recruitment and assessments",
    icon: Users,
    category: "people",
    baseRoute: "/hiring",
    requiredPermission: "can_view_hiring",
    modules: [
      { id: "dashboard", name: "Dashboard", description: "Hiring overview and metrics", route: "/dashboard" },
      { id: "candidates", name: "Candidates", description: "Manage candidates", route: "/candidates" },
      { id: "assessments", name: "Assessments", description: "Technical assessments", route: "/assessments" },
      { id: "questions", name: "Question Bank", description: "Assessment questions library", route: "/questions" },
      { id: "templates", name: "Templates", description: "Assessment templates", route: "/templates" },
      { id: "analytics", name: "Analytics", description: "Hiring analytics and reports", route: "/analytics" },
    ],
  },
  learning: {
    id: "learning",
    name: "Learning",
    description: "Learning paths and courses",
    icon: GraduationCap,
    category: "people",
    baseRoute: "/learning",
    requiredPermission: "can_view_learning",
    modules: [],
  },
  crm: {
    id: "crm",
    name: "CRM",
    description: "Customer relationship management",
    icon: Building2,
    category: "business",
    baseRoute: "/crm",
    requiredPermission: "can_view_crm",
    modules: [
      { id: "overview", name: "Overview", description: "CRM dashboard and pipeline", route: "" },
      { id: "inbox", name: "Inbox", description: "Email inbox and communications", route: "/inbox" },
      { id: "agents", name: "AI Agents", description: "Configure AI sales agents", route: "/agents" },
      { id: "activities", name: "Activities", description: "Activity tracking and logs", route: "/activities" },
      { id: "automations", name: "Automations", description: "Sales automations and sequences", route: "/automations" },
      { id: "calendar", name: "Calendar", description: "Meeting and event calendar", route: "/calendar" },
    ],
  },
  email_marketing: {
    id: "email_marketing",
    name: "Email Marketing",
    description: "Email campaigns and automation",
    icon: Mail,
    category: "business",
    baseRoute: "/email-marketing",
    requiredPermission: "can_view_crm",
    modules: [
      { id: "campaigns", name: "Campaigns", description: "Email campaign management", route: "/campaigns" },
      { id: "templates", name: "Templates", description: "Email templates library", route: "/templates" },
      { id: "settings", name: "Settings", description: "Email settings and domains", route: "/settings" },
    ],
  },
  docs: {
    id: "docs",
    name: "Docs",
    description: "Documentation and wiki",
    icon: FileText,
    category: "productivity",
    baseRoute: "/docs",
    requiredPermission: "can_view_docs",
    modules: [],
  },
  forms: {
    id: "forms",
    name: "Forms",
    description: "Form builder and submissions",
    icon: ClipboardList,
    category: "productivity",
    baseRoute: "/forms",
    requiredPermission: "can_view_forms",
    modules: [],
  },
  oncall: {
    id: "oncall",
    name: "On-Call",
    description: "On-call schedules and rotations",
    icon: Phone,
    category: "engineering",
    baseRoute: "/oncall",
    requiredPermission: "can_view_oncall",
    modules: [],
  },
  booking: {
    id: "booking",
    name: "Booking",
    description: "Calendar booking and scheduling",
    icon: CalendarCheck,
    category: "business",
    baseRoute: "/booking",
    requiredPermission: "can_view_booking",
    modules: [
      { id: "event_types", name: "Event Types", description: "Manage bookable event types", route: "/event-types" },
      { id: "availability", name: "Availability", description: "Set your availability schedule", route: "/availability" },
      { id: "calendars", name: "Calendars", description: "Connect external calendars", route: "/calendars" },
    ],
  },
  uptime: {
    id: "uptime",
    name: "Uptime",
    description: "Endpoint monitoring and incident management",
    icon: MonitorCheck,
    category: "engineering",
    baseRoute: "/uptime",
    requiredPermission: "can_view_uptime",
    modules: [
      { id: "monitors", name: "Monitors", description: "HTTP, TCP, and WebSocket endpoint monitors", route: "/monitors" },
      { id: "incidents", name: "Incidents", description: "Active and resolved incidents", route: "/incidents" },
      { id: "history", name: "History", description: "Check history and uptime reports", route: "/history" },
    ],
  },
  automations: {
    id: "automations",
    name: "Automations",
    description: "Platform-wide workflow automations",
    icon: Zap,
    category: "productivity",
    baseRoute: "/automations",
    requiredPermission: "can_view_automations",
    modules: [],
  },
  agents: {
    id: "agents",
    name: "AI Agents",
    description: "AI-powered automation agents",
    icon: Bot,
    category: "productivity",
    baseRoute: "/agents",
    requiredPermission: "can_view_agents",
    modules: [],
  },
  insights: {
    id: "insights",
    name: "Insights",
    description: "Developer productivity metrics and team analytics",
    icon: TrendingUp,
    category: "engineering",
    baseRoute: "/insights",
    requiredPermission: "can_view_insights",
    modules: [
      { id: "team_overview", name: "Team Overview", description: "Team-wide velocity, efficiency, and workload distribution", route: "" },
      { id: "leaderboard", name: "Leaderboard", description: "Ranked developer metrics", route: "/leaderboard" },
      { id: "developer_drilldown", name: "Developer Drill-down", description: "Individual developer metrics deep-dive", route: "/developers" },
    ],
  },
};

// Get app definition by ID
export function getAppById(appId: string): AppDefinition | undefined {
  return APP_CATALOG[appId];
}

// Get all apps as array
export function getAllApps(): AppDefinition[] {
  return Object.values(APP_CATALOG);
}

// Get apps by category
export function getAppsByCategory(category: AppCategory): AppDefinition[] {
  return Object.values(APP_CATALOG).filter((app) => app.category === category);
}

// Check if a route belongs to an app
export function getAppForRoute(pathname: string): AppDefinition | undefined {
  for (const app of Object.values(APP_CATALOG)) {
    if (pathname === app.baseRoute || pathname.startsWith(`${app.baseRoute}/`)) {
      return app;
    }
  }
  return undefined;
}

// Check if a route belongs to a specific module
export function getModuleForRoute(
  pathname: string
): { app: AppDefinition; module: AppModule } | undefined {
  for (const app of Object.values(APP_CATALOG)) {
    if (pathname === app.baseRoute || pathname.startsWith(`${app.baseRoute}/`)) {
      const relativePath = pathname.replace(app.baseRoute, "");
      for (const module of app.modules) {
        if (relativePath === module.route || relativePath.startsWith(`${module.route}/`)) {
          return { app, module };
        }
      }
      // If no module matched but app matched, return app without module
      return undefined;
    }
  }
  return undefined;
}

// System app bundle templates
export const SYSTEM_BUNDLES: AppBundleTemplate[] = [
  {
    id: "engineering",
    name: "Engineering",
    description: "Apps for software development teams",
    icon: "Code",
    color: "#2563eb",
    isSystem: true,
    appConfig: {
      dashboard: { enabled: true },
      tracking: { enabled: true, modules: { standups: true, blockers: true, time: true } },
      sprints: { enabled: true, modules: { board: true, epics: true, tasks: true, backlog: true } },
      tickets: { enabled: true },
      docs: { enabled: true },
      learning: { enabled: true },
      oncall: { enabled: true },
      uptime: { enabled: true, modules: { monitors: true, incidents: true, history: true } },
      reviews: { enabled: false },
      hiring: { enabled: false },
      crm: { enabled: false },
      email_marketing: { enabled: false },
      forms: { enabled: false },
      booking: { enabled: false },
      automations: { enabled: true },
      agents: { enabled: true },
      insights: { enabled: false },
    },
  },
  {
    id: "people",
    name: "People",
    description: "Apps for HR and people operations",
    icon: "Heart",
    color: "#f43f5e",
    isSystem: true,
    appConfig: {
      dashboard: { enabled: true },
      reviews: { enabled: true, modules: { cycles: true, goals: true, peer_requests: true, manage: true } },
      hiring: {
        enabled: true,
        modules: { dashboard: true, candidates: true, assessments: true, questions: true, templates: true, analytics: true },
      },
      learning: { enabled: true },
      docs: { enabled: true },
      forms: { enabled: true },
      tracking: { enabled: false },
      sprints: { enabled: false },
      tickets: { enabled: false },
      crm: { enabled: false },
      email_marketing: { enabled: false },
      oncall: { enabled: false },
      uptime: { enabled: false },
      booking: { enabled: false },
      automations: { enabled: true },
      agents: { enabled: true },
      insights: { enabled: false },
    },
  },
  {
    id: "business",
    name: "Business",
    description: "Apps for sales and customer success",
    icon: "Briefcase",
    color: "#06b6d4",
    isSystem: true,
    appConfig: {
      dashboard: { enabled: true },
      crm: {
        enabled: true,
        modules: { overview: true, inbox: true, agents: true, activities: true, automations: true, calendar: true },
      },
      email_marketing: { enabled: true, modules: { campaigns: true, templates: true, settings: true } },
      tickets: { enabled: true },
      docs: { enabled: true },
      forms: { enabled: true },
      tracking: { enabled: false },
      sprints: { enabled: false },
      reviews: { enabled: false },
      hiring: { enabled: false },
      learning: { enabled: false },
      oncall: { enabled: false },
      uptime: { enabled: false },
      booking: { enabled: true, modules: { event_types: true, availability: true, calendars: true } },
      automations: { enabled: true },
      agents: { enabled: true },
      insights: { enabled: false },
    },
  },
  {
    id: "full_access",
    name: "Full Access",
    description: "Access to all apps and modules",
    icon: "Shield",
    color: "#9333ea",
    isSystem: true,
    appConfig: {
      dashboard: { enabled: true },
      tracking: { enabled: true, modules: { standups: true, blockers: true, time: true } },
      sprints: { enabled: true, modules: { board: true, epics: true, tasks: true, backlog: true } },
      tickets: { enabled: true },
      reviews: { enabled: true, modules: { cycles: true, goals: true, peer_requests: true, manage: true } },
      hiring: {
        enabled: true,
        modules: { dashboard: true, candidates: true, assessments: true, questions: true, templates: true, analytics: true },
      },
      learning: { enabled: true },
      crm: {
        enabled: true,
        modules: { overview: true, inbox: true, agents: true, activities: true, automations: true, calendar: true },
      },
      email_marketing: { enabled: true, modules: { campaigns: true, templates: true, settings: true } },
      docs: { enabled: true },
      forms: { enabled: true },
      oncall: { enabled: true },
      uptime: { enabled: true, modules: { monitors: true, incidents: true, history: true } },
      booking: { enabled: true, modules: { event_types: true, availability: true, calendars: true } },
      automations: { enabled: true },
      agents: { enabled: true },
      insights: { enabled: true, modules: { team_overview: true, leaderboard: true, developer_drilldown: true } },
    },
  },
];

// Get bundle by ID
export function getBundleById(bundleId: string): AppBundleTemplate | undefined {
  return SYSTEM_BUNDLES.find((b) => b.id === bundleId);
}

// Map sidebar items to app IDs
export const SIDEBAR_TO_APP_MAP: Record<string, string> = {
  "/dashboard": "dashboard",
  "/tracking": "tracking",
  "/tracking/standups": "tracking",
  "/tracking/blockers": "tracking",
  "/tracking/time": "tracking",
  "/sprints": "sprints",
  "/tickets": "tickets",
  "/reviews": "reviews",
  "/reviews/cycles": "reviews",
  "/reviews/goals": "reviews",
  "/reviews/peer-requests": "reviews",
  "/reviews/manage": "reviews",
  "/hiring": "hiring",
  "/hiring/dashboard": "hiring",
  "/hiring/candidates": "hiring",
  "/hiring/assessments": "hiring",
  "/hiring/questions": "hiring",
  "/hiring/templates": "hiring",
  "/hiring/analytics": "hiring",
  "/learning": "learning",
  "/crm": "crm",
  "/crm/inbox": "crm",
  "/crm/activities": "crm",
  "/crm/calendar": "crm",
  "/automations": "automations",
  "/automations/new": "automations",
  "/agents": "agents",
  "/agents/new": "agents",
  "/email-marketing": "email_marketing",
  "/email-marketing/campaigns": "email_marketing",
  "/email-marketing/templates": "email_marketing",
  "/email-marketing/settings": "email_marketing",
  "/docs": "docs",
  "/forms": "forms",
  "/oncall": "oncall",
  "/booking": "booking",
  "/booking/event-types": "booking",
  "/booking/availability": "booking",
  "/booking/calendars": "booking",
  "/uptime": "uptime",
  "/uptime/monitors": "uptime",
  "/uptime/incidents": "uptime",
  "/uptime/history": "uptime",
  "/insights": "insights",
  "/insights/leaderboard": "insights",
  "/insights/developers": "insights",
};

// Get app ID from pathname
export function getAppIdFromPath(pathname: string): string | undefined {
  // Check exact match first
  if (SIDEBAR_TO_APP_MAP[pathname]) {
    return SIDEBAR_TO_APP_MAP[pathname];
  }

  // Check prefix matches
  for (const [route, appId] of Object.entries(SIDEBAR_TO_APP_MAP)) {
    if (pathname.startsWith(route + "/")) {
      return appId;
    }
  }

  // Fallback to getAppForRoute
  const app = getAppForRoute(pathname);
  return app?.id;
}
